lang/wasm_instance_cache.hh

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include "cql3/functions/function_name.hh"
+#include "utils/overloaded_functor.hh"
+#include "wasmtime.hh"
+#include "lang/wasm.hh"
+#include <exception>
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/core/scheduling.hh>
+#include "utils/loading_cache.hh"
+#include "utils/hash.hh"
+#include "utils/overloaded_functor.hh"
+#include <chrono>
+#include <cstdint>
+#include <queue>
+#include <unordered_map>
+
+extern logging::logger wasm_logger;
+
+namespace wasm {
+
+struct udf_instance {
+    wasmtime::Store store;
+    wasmtime::Instance instance;
+    wasmtime::Func func;
+    wasmtime::Memory memory;
+};
+
+// For each UDF full name and a scheduling group, we store a future that is ready
+// when no other task uses the same UDF instance. When an instance is retrieved
+// from the cache, the future is replaced with a new one that becomes ready when
+// the instance is returned. To avoid conflicts between tasks, each time an 
+// instance is requested, the future associated with it is removed from the cache
+// and a new one is added. When the current instance user is done with it, the
+// future that was associated with it is becomes ready.
+// This way, an instance stays in the cache as long as it isn't used by any task
+// and if there are tasks that want to use it, they get it immediately after
+// the previous user is finished.
+class instance_cache {
+public:
+    struct stats {
+        uint64_t cache_hits = 0;
+        uint64_t cache_misses = 0;
+        uint64_t cache_blocks = 0;
+    };
+
+    static stats& shard_stats() {
```

Comment:
let's use seastar::metrics for that, just like the rest of our code base, here's a working example for convenience: https://github.com/scylladb/scylla/blob/42f51b2f7b65f9f52c94bbcbac5668ebf769ea9d/db/view/view.cc#L216-L231 Then we get them exported and visible in Grafana for free, and they're already per-shard by design.